File: plugins/albino/mcp-servers/project-memory/src/embedding.ts
/**
 * Neural embedding using @huggingface/transformers.
 *
 * Generates dense 384-dimensional float32 vectors via the
 * Xenova/all-MiniLM-L6-v2 sentence-transformer model. The pipeline is
 * lazy-initialised once on first call and reused for subsequent calls.
 */

import { pipeline, type FeatureExtractionPipeline } from "@huggingface/transformers";

/** Dimensionality of the all-MiniLM-L6-v2 model output. */
export const EMBEDDING_DIM = 384;

/** The model used for embedding. */
const MODEL_ID = "Xenova/all-MiniLM-L6-v2";

let pipelineInstance: FeatureExtractionPipeline | null = null;

async function getPipeline(): Promise<FeatureExtractionPipeline> {
  if (!pipelineInstance) {
    pipelineInstance = (await pipeline("feature-extraction", MODEL_ID, {
      dtype: "fp32"
    })) as FeatureExtractionPipeline;
  }
  return pipelineInstance;
}

/**
 * Generate dense embeddings for an array of texts.
 * Returns one float32 vector per input text, each of length EMBEDDING_DIM (384).
 */
export async function embed(texts: string[]): Promise<number[][]> {
  if (texts.length === 0) {
    return [];
  }

  const extractor = await getPipeline();
  const output = await extractor(texts, { pooling: "mean", normalize: true });

  // output.tolist() returns number[][] — one row per input text
  const nested = output.tolist() as number[][];
  return nested;
}

/** Build the text that gets embedded for a memory record. */
export function memoryEmbedText(content: string, summary: string | null, tags: string[]): string {
  const parts: string[] = [content];
  if (summary) {
    parts.push(summary);
  }

  if (tags.length > 0) {
    parts.push(tags.join(" "));
  }

  return parts.join(" ");
}

/** Compute cosine similarity between two dense vectors. Range [-1, 1]. */
export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length || a.length === 0) {
    return 0;
  }

  let dot = 0;
  let magA = 0;
  let magB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i]! * b[i]!;
    magA += a[i]! * a[i]!;
    magB += b[i]! * b[i]!;
  }

  const denom = Math.sqrt(magA) * Math.sqrt(magB);
  if (denom === 0) {
    return 0;
  }

  return dot / denom;
}

File: plugins/albino/mcp-servers/project-memory/src/project_memory/embedding.py
import asyncio
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from fastembed import TextEmbedding

EMBEDDING_DIM = 384
_MODEL_ID = "BAAI/bge-small-en-v1.5"

_model: "TextEmbedding | None" = None


def _get_model() -> "TextEmbedding":
    global _model
    if _model is None:
        from fastembed import TextEmbedding

        _model = TextEmbedding(_MODEL_ID)
    return _model


def embed_sync(texts: list[str]) -> list[list[float]]:
    if not texts:
        return []
    model = _get_model()
    return [emb.tolist() for emb in model.embed(texts)]


async def embed(texts: list[str]) -> list[list[float]]:
    if not texts:
        return []
    return await asyncio.to_thread(embed_sync, texts)


def memory_embed_text(content: str, summary: str | None, tags: list[str]) -> str:
    parts = [content]
    if summary:
        parts.append(summary)
    if tags:
        parts.append(" ".join(tags))
    return " ".join(parts)


def cosine_similarity(a: list[float], b: list[float]) -> float:
    if len(a) != len(b) or not a:
        return 0.0
    dot = sum(x * y for x, y in zip(a, b))
    mag_a = sum(x * x for x in a) ** 0.5
    mag_b = sum(x * x for x in b) ** 0.5
    denom = mag_a * mag_b
    return dot / denom if denom != 0 else 0.0
